nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/CopyS3Object.java

```diff
@@ -0,0 +1,145 @@
+            .fromPropertyDescriptor(KEY)
+            .name("copy-s3-target-key")
+            .displayName("Target Key")
+            .description("The target key in the target bucket")
+            .defaultValue("")
+            .build();
+
+    public static final List<PropertyDescriptor> properties = List.of(
+            SOURCE_BUCKET,
+            SOURCE_KEY,
+            TARGET_BUCKET,
+            TARGET_KEY,
+            AWS_CREDENTIALS_PROVIDER_SERVICE,
+            S3_REGION,
+            TIMEOUT,
+            FULL_CONTROL_USER_LIST,
+            READ_USER_LIST,
+            WRITE_USER_LIST,
+            READ_ACL_LIST,
+            WRITE_ACL_LIST,
+            CANNED_ACL,
+            OWNER,
+            SSL_CONTEXT_SERVICE,
+            ENDPOINT_OVERRIDE,
+            SIGNER_OVERRIDE,
+            S3_CUSTOM_SIGNER_CLASS_NAME,
+            S3_CUSTOM_SIGNER_MODULE_LOCATION,
+            PROXY_CONFIGURATION_SERVICE);
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final AmazonS3Client s3;
+        try {
+            s3 = getS3Client(context, flowFile.getAttributes());
+        } catch (Exception e) {
+            getLogger().error("Failed to initialize S3 client", e);
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        final String sourceBucket = context.getProperty(SOURCE_BUCKET).evaluateAttributeExpressions(flowFile).getValue();
+        final String sourceKey = context.getProperty(SOURCE_KEY).evaluateAttributeExpressions(flowFile).getValue();
+        final String targetBucket = context.getProperty(TARGET_BUCKET).evaluateAttributeExpressions(flowFile).getValue();
+        final String targetKey = context.getProperty(TARGET_KEY).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            CopyObjectRequest request = new CopyObjectRequest(sourceBucket, sourceKey, targetBucket, targetKey);
+            AccessControlList acl = createACL(context, flowFile);
+            if (acl != null) {
+                request.setAccessControlList(acl);
+            }
+            CannedAccessControlList cannedAccessControlList = createCannedACL(context, flowFile);
+
+            if (cannedAccessControlList != null) {
+                request.setCannedAccessControlList(cannedAccessControlList);
+            }
+
+            s3.copyObject(request);
+            session.getProvenanceReporter().send(flowFile, getTransitUrl(targetBucket, targetKey));
+
+            session.transfer(flowFile, REL_SUCCESS);
+        } catch (Exception ex) {
+            getLogger().error("Copy S3 Object Request failed with error:", ex);
```

Comment:
Are the exception details you mention a stack trace or a generally user beneficial string describing what went wrong?  We need to avoid large attributes being carried around - they're not designed for that.

If we're going to put error related attributes on we also need to be thoughtful about clearing them out when the success path occurs for the same component that would have put them on.

Reply:
Thanks, @joewitt for raising valid points to consider. I feel the attributes are beneficial and relatively small with no stack traces as you can inspect in the code below.
FYI, the method I suggested is used in all 5 of the existing S3Object related processors. 
@joewitt Are you OK with the new processors using the extractExceptionDetails() like the existing S3 processors do?

This is the method that adds exception related attributes.
`protected FlowFile extractExceptionDetails(final Exception e, final ProcessSession session, FlowFile flowFile) {
        flowFile = session.putAttribute(flowFile, "s3.exception", e.getClass().getName());
        if (e instanceof AmazonS3Exception) {
            flowFile = putAttribute(session, flowFile, "s3.additionalDetails", ((AmazonS3Exception) e).getAdditionalDetails());
        }
        if (e instanceof final AmazonServiceException ase) {
            flowFile = putAttribute(session, flowFile, "s3.statusCode", ase.getStatusCode());
            flowFile = putAttribute(session, flowFile, "s3.errorCode", ase.getErrorCode());
            flowFile = putAttribute(session, flowFile, "s3.errorMessage", ase.getErrorMessage());
        }
        return flowFile;
    }
`

Reply:
@joewitt I will add a new ticket for AbstractS3Processor class to clear out exception related attributes that it writes in its extractExceptionDetails() method to resolve your point about "when the success path occurs for the same component that would have put them on."

Reply:
thanks.  In general we have uses of attributes that are increasingly specific to a point in time in the flow/a particular component.  While this pattern can be useful for giving users some added context and not having dive into logs to find it we must also then be equally motivated to clean these values up when they're no longer relevant

Reply:
As for being ok with 'extract exception details' I am not familiar with precisely what is in the output.  Provided what it is would be a rather short and meant for more like end user reading then sure.  If it is really developer level info then this is often not helpful to end users and/or can be far too large.